projects/plugins/crm/src/automation/commons/conditions/contacts/class-contact-tag.php

```diff
@@ -65,80 +64,44 @@ private function has_tag_by_name( $data, $tag_name ) {
 		return false;
 	}
 
-	/**
-	 * Check for valid parameters.
-	 *
-	 * @since $$next-version$$
-	 *
-	 * @param string $operator The operator.
-	 * @param mixed  $data The data to validate.
-	 * @param mixed  $previous_data The previous data to validate.
-	 * @return bool True if parameters are valid, false otherwise.
-	 */
-	private function check_for_valid_parameters( $operator, $data, $previous_data ) {
-		switch ( $operator ) {
-			case 'tag_added':
-			case 'tag_removed':
-				if ( $previous_data === null || ! $this->is_valid_contact_tag_data( $data ) || ! $this->is_valid_contact_tag_data( $previous_data ) ) {
-					$this->logger->log( 'Invalid contact tag data' );
-
-					return false;
-				}
-				break;
-			case 'has_tag':
-				if ( ! $this->is_valid_contact_tag_data( $data ) ) {
-					$this->logger->log( 'Invalid contact tag data' );
-					$this->condition_met = false;
-
-					return false;
-				}
-				break;
-			default:
-				$this->logger->log( 'Unknown operator: ' . $operator );
-				return false;
-		}
-
-		return true;
-	}
-
 	/**
 	 * Executes the condition. If the condition is met, the value stored in the
 	 * attribute $condition_met is set to true; otherwise, it is set to false.
 	 *
 	 * @since $$next-version$$
 	 *
-	 * @param mixed  $data Data passed from the trigger.
-	 * @param ?mixed $previous_data (Optional) The data before being changed.
+	 * @param Data_Type $data Data passed from the trigger.
 	 * @return void
 	 *
 	 * @throws Automation_Exception If an invalid operator is encountered.
 	 */
-	public function execute( $data, $previous_data = null ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+	public function execute( Data_Type $data ) {
+
+		/** @var Contact $contact */
+		$contact = $data->get_data();
+
 		$operator = $this->get_attributes()['operator'];
 		$tag      = $this->get_attributes()['tag'];
 
 		$this->check_for_valid_operator( $operator );
 
-		if ( ! $this->check_for_valid_parameters( $operator, $data, $previous_data ) ) {
-			$this->condition_met = false;
-			return;
-		}
+		$previous_data = $data->get_previous_data();
```

Comment:
We should add a guard clause to bail early if "previous data" isn't a `Contact` instance.
The current implementation could end in a fatal error when `$previous_data` is passed to `$this->has_tag_by_name()`

```suggestion
		$previous_data = $data->get_previous_data();
		
		if ( ! $previous_data instanceof Contact ) {
			$this->condition_met = false;

			return;
		}
```

Reply:
Correction: we do not use `$previous_state` in `has_tag`, so we cannot add the check where I suggested, but we still need to check further down.

Reply:
`previous_data` is validated in the `Data_Type_Base` constructor, calling to the `validate_data` method, that is implemented by each `Data_Type`.